Sources/SwiftFormatCore/DocumentationComment.swift

```diff
@@ -0,0 +1,339 @@
+  public var briefSummary: Paragraph? = nil
+
+  /// A collection of otherwise uncategorized body nodes at the top level of the comment text.
+  ///
+  /// If a brief summary paragraph was extracted from the comment, it will not be present in this
+  /// collection.
+  public var bodyNodes: [Markup] = []
+
+  /// The structural layout of the parameter descriptions in the comment.
+  public var parameterLayout: ParameterLayout? = nil
+
+  /// Descriptions of parameters to a function, if any.
+  public var parameters: [Parameter] = []
+
+  /// A description of the return value of a function.
+  ///
+  /// If present, this value is a copy of the `Paragraph` node from the comment but with the
+  /// `Returns:` prefix removed for convenience.
+  public var returns: Paragraph? = nil
+
+  /// A description of an error thrown by a function.
+  ///
+  /// If present, this value is a copy of the `Paragraph` node from the comment but with the
+  /// `Returns:` prefix removed for convenience.
+  public var `throws`: Paragraph? = nil
+
+  /// Creates a new `DocumentationComment` with information extracted from the leading trivia of the
+  /// given syntax node.
+  ///
+  /// If the syntax node does not have a preceding documentation comment, this initializer returns
+  /// `nil`.
+  ///
+  /// - Parameter node: The syntax node from which the documentation comment should be extracted.
+  public init?<Node: SyntaxProtocol>(extractedFrom node: Node) {
+    guard let commentInfo = documentationCommentText(extractedFrom: node.leadingTrivia) else {
+      return nil
+    }
+
+    // Disable smart quotes and dash conversion since we want to preserve the original content of
+    // the comments instead of doing documentation generation.
+    let doc = Document(parsing: commentInfo.text, options: [.disableSmartOpts])
+    self.init(markup: doc)
+  }
+
+  /// Creates a new `DocumentationComment` from the given `Markup` node.
+  private init(markup: Markup) {
+    // Extract the first paragraph as the brief summary. It will *not* be included in the body
+    // nodes.
+    let remainingChildren: DropFirstSequence<MarkupChildren>
+    if let firstParagraph = markup.child(through: [(0, Paragraph.self)]) {
+      briefSummary = firstParagraph.detachedFromParent as? Paragraph
+      remainingChildren = markup.children.dropFirst()
+    } else {
+      briefSummary = nil
+      remainingChildren = markup.children.dropFirst(0)
+    }
+
+    for child in remainingChildren {
+      if var list = child.detachedFromParent as? UnorderedList {
+        // An unordered list could be one of the following:
+        //
+        // 1.  A parameter outline:
+        //     - Parameters:
+        //       - x: ...
+        //       - y: ...
+        //
+        // 2.  An exploded parameter list:
+        //     - Parameter x: ...
+        //     - Parameter y: ...
+        //
+        // 3.  Some other simple field, like `Returns:`.
+        extractParameterOutline(from: &list)
```

Comment:
It might be worth somehow capturing that the order here is relevant for setting `parameterLayout` correctly.

Reply:
Good idea; done.